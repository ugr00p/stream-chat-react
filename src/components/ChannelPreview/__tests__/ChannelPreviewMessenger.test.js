import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import axios from 'axios';
import { render, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

import {
  useMockedApis,
  generateUser,
  generateChannel,
  getTestClientWithUser,
  getOrCreateChannelApi,
} from 'mock-builders';

import ChannelPreviewMessenger from '../ChannelPreviewMessenger';

jest.mock('axios');

describe('ChannelPreviewMessenger', () => {
  const clientUser = generateUser();
  let chatClient;
  let channel;
  const renderComponent = (props) => {
    return (
      <ChannelPreviewMessenger
        channel={channel}
        latestMessage="This is latest message !!!"
        unread={10}
        latestMessageLength={6}
        displayTitle="Channel name"
        displayImage="https://randomimage.com/src.jpg"
        setActiveChannel={jest.fn()}
        {...props}
      />
    );
  };

  const initializeChannel = async (c) => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useMockedApis(axios, [getOrCreateChannelApi(c)]);

    channel = chatClient.channel('messaging');

    await channel.watch();
  };

  beforeEach(async () => {
    chatClient = await getTestClientWithUser(clientUser);
    await initializeChannel(generateChannel());
  });

  it('should render correctly', () => {
    const tree = renderer.create(renderComponent()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call setActiveChannel on click', async () => {
    const setActiveChannel = jest.fn();
    const { getByTestId } = render(
      renderComponent({
        setActiveChannel,
        watchers: {},
      }),
    );

    await waitFor(() => {
      expect(getByTestId('channel-preview-button')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('channel-preview-button'));

    await waitFor(() => {
      // eslint-disable-next-line jest/prefer-called-with
      expect(setActiveChannel).toHaveBeenCalledTimes(1);
      expect(setActiveChannel).toHaveBeenCalledWith(channel, {});
    });
  });
});
