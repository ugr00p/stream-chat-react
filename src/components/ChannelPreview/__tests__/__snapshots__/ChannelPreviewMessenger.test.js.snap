// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelPreviewMessenger should render correctly 1`] = `
<button
  className="str-chat__channel-preview-messenger str-chat__channel-preview-messenger--unread "
  data-testid="channel-preview-button"
  onClick={[Function]}
>
  <div
    className="str-chat__channel-preview-messenger--left"
  >
    <div
      className="str-chat__avatar str-chat__avatar--circle"
      data-testid="avatar"
      onClick={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "flexBasis": 40,
          "fontSize": 20,
          "height": 40,
          "lineHeight": "40px",
          "width": 40,
        }
      }
    >
      <img
        alt=""
        className="str-chat__avatar-image"
        data-testid="avatar-img"
        onError={[Function]}
        onLoad={[Function]}
        src="https://randomimage.com/src.jpg"
        style={
          Object {
            "flexBasis": 40,
            "height": 40,
            "objectFit": "cover",
            "width": 40,
          }
        }
      />
    </div>
  </div>
  <div
    className="str-chat__channel-preview-messenger--right"
  >
    <div
      className="str-chat__channel-preview-messenger--name"
    >
      <span>
        Channel name
      </span>
    </div>
    <div
      className="str-chat__channel-preview-messenger--last-message"
    >
      Thi...
    </div>
  </div>
</button>
`;
